### PR TITLE
Floor and ceil methods during pandas.eval which are provided by numexpr

### LINKED ISSUE #24353: Need support for floor and ceil methods during pandas.eval which are provided by numexpr

### PR DESCRIPTION
**Pull request by anjsudh**
- [x] closes #24353
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: doc/source/whatsnew/v0.24.0.rst [hunk 1]
```diff
@@ -372,6 +372,7 @@ Other Enhancements
 - :meth:`DataFrame.to_stata` and :class:`pandas.io.stata.StataWriter117` can write mixed sting columns to Stata strl format (:issue:`23633`)
 - :meth:`DataFrame.between_time` and :meth:`DataFrame.at_time` have gained the an ``axis`` parameter (:issue:`8839`)
 - :class:`IntervalIndex` has gained the :attr:`~IntervalIndex.is_overlapping` attribute to indicate if the ``IntervalIndex`` contains any overlapping intervals (:issue:`23309`)
+- :meth:`eval()` now supports use of `floor` and `ceil` in the expression
```
Comment (TomAugspurger) [inline doc/source/whatsnew/v0.24.0.rst]: Add the issue number here. This will also work in DataFrame.query, right?

**Review by TomAugspurger — COMMENTED** [review]
Needs tests too :)

File: doc/source/whatsnew/v0.24.0.rst (hunk 1 above)
Comment (jreback) [reply doc/source/whatsnew/v0.24.0.rst]: move this underneath this note:

``- Added log10 to the list of supported functions in :meth:`DataFrame.eval` (:issue:`24139`)``

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/computation/ops.py [hunk 2]
```diff
@@ -541,7 +542,7 @@ def __unicode__(self):
 class FuncNode(object):
 
     def __init__(self, name):
-        if name not in _mathops:
+        if name not in _mathops or not hasattr(np, name) :
```
Comment (jreback) [inline pandas/core/computation/ops.py]: why is this change needed?

File: pandas/tests/computation/test_eval.py [hunk 3]
```diff
@@ -1627,10 +1627,21 @@ def test_unary_functions(self):
         a = df.a
         for fn in self.unary_fns:
             expr = "{0}(a)".format(fn)
-            got = self.eval(expr)
             with np.errstate(all='ignore'):
-                expect = getattr(np, fn)(a)
-            tm.assert_series_equal(got, expect, check_names=False)
+                if hasattr(np, fn):
```
Comment (jreback) [inline pandas/tests/computation/test_eval.py]: why this change?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/computation/ops.py (hunk 2 above)
Comment (anjsudh) [reply pandas/core/computation/ops.py]: so older numpy does not support floor and ceil methods. They were introduced only in the 1.15. So, i felt it would be more appropriate to check if its supported instead of having a version check.

File: pandas/core/computation/check.py [hunk 4]
```diff
@@ -3,11 +3,13 @@
 
 _NUMEXPR_INSTALLED = False
 _MIN_NUMEXPR_VERSION = "2.6.1"
+_ne_version_under_2_6_9 = None
 
 try:
     import numexpr as ne
     ver = LooseVersion(ne.__version__)
     _NUMEXPR_INSTALLED = ver >= LooseVersion(_MIN_NUMEXPR_VERSION)
+    _ne_version_under_2_6_9 = ver < LooseVersion('2.6.9')
```
Comment (jreback) [inline pandas/core/computation/check.py]: you don't need to add this here

File: pandas/core/computation/expressions.py [hunk 5]
```diff
@@ -20,6 +20,7 @@
 _TEST_MODE = None
 _TEST_RESULT = None
 _USE_NUMEXPR = _NUMEXPR_INSTALLED
+_ne_version_under2_6_9 = _ne_version_under_2_6_9
```
Comment (jreback) [inline pandas/core/computation/expressions.py]: you don't need to add this here

File: pandas/core/computation/check.py (hunk 4 above)
Comment (jreback) [reply pandas/core/computation/check.py]: rather export _NUMEXPR_VERSION as well

File: pandas/core/computation/ops.py [hunk 6]
```diff
@@ -23,8 +25,10 @@
 
 _unary_math_ops = ('sin', 'cos', 'exp', 'log', 'expm1', 'log1p',
                    'sqrt', 'sinh', 'cosh', 'tanh', 'arcsin', 'arccos',
-                   'arctan', 'arccosh', 'arcsinh', 'arctanh', 'abs', 'log10')
+                   'arctan', 'arccosh', 'arcsinh', 'arctanh', 'abs', 'log10',
+                   'floor', 'ceil')
```
Comment (jreback) [inline pandas/core/computation/ops.py]: just do this
```
_unary_math_ops = [...]
if _NUMEXPR_INSTALLED and _NUMEXPR_VERSION >= LooseVersion('2.6.9'):
    _unary_math_ops += ['floor', 'ceil']
```

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/computation/test_eval.py [hunk 7]
```diff
@@ -1625,7 +1641,26 @@ def eval(self, *args, **kwargs):
     def test_unary_functions(self):
         df = DataFrame({'a': np.random.randn(10)})
         a = df.a
-        for fn in self.unary_fns:
+        unary_functions = [x for x in self.unary_fns
```
Comment (jreback) [inline pandas/tests/computation/test_eval.py]: can you parameterize this over the unary functions (it will automatically just work according to the version of numexpr)

File: pandas/tests/computation/test_eval.py [hunk 8]
```diff
@@ -1625,7 +1641,26 @@ def eval(self, *args, **kwargs):
...
+                           if x not in ('floor', 'ceil')]
+        for fn in unary_functions:
+            expr = "{0}(a)".format(fn)
+            got = self.eval(expr)
+            with np.errstate(all='ignore'):
+                expect = getattr(np, fn)(a)
+            tm.assert_series_equal(got, expect, check_names=False)
+
+    def test_floor_and_ceil_functions_raise_error(self, ne_lt_2_6_9):
```
Comment (jreback) [inline pandas/tests/computation/test_eval.py]: can you parametrize

File: pandas/tests/computation/test_eval.py [hunk 9]
```diff
@@ -1625,7 +1641,26 @@ def eval(self, *args, **kwargs):
...
+        for fn in ('floor', 'ceil'):
+            msg = "\"{0}\" is not a supported function".format(fn)
+            with pytest.raises(ValueError, match=msg):
+                expr = "{0}(100)".format(fn)
+                self.eval(expr)
+
+    def test_floor_and_ceil_functions_evaluate_expressions(self, ne_gt_2_6_9):
```
Comment (jreback) [inline pandas/tests/computation/test_eval.py]: can you parametetrize (note i don't think this test adds anything beyond the basic one)

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by anjsudh** [comment]
Getting an error: "err: pandas should not import: numexpr". I haven't imported numexpr in my diff. The error resolves when I don't import _NUMEXPR_INSTALLED and _NUMEXPR_VERSION. Any ideas on how to go about the version check without the import?

Also, is there a reason for us to check that numexpr is not imported? The import is anyways in a try catch block

Any inputs would help here?

File: pandas/core/computation/ops.py [hunk 10]
```diff
@@ -14,7 +15,9 @@
 import pandas as pd
 from pandas.core.base import StringMixin
 import pandas.core.common as com
+from pandas.core.computation.check import _NUMEXPR_VERSION
 from pandas.core.computation.common import _ensure_decoded, _result_type_many
+from pandas.core.computation.expressions import _NUMEXPR_INSTALLED
```
Comment (jreback) [inline pandas/core/computation/ops.py]: import from check

File: pandas/tests/computation/test_eval.py [hunk 11]
```diff
@@ -31,6 +32,8 @@
                                  assert_numpy_array_equal, assert_series_equal,
                                  assert_produces_warning)
 from pandas.compat import PY3, reduce
+from pandas.core.computation.check import _NUMEXPR_VERSION
```
Comment (jreback) [inline pandas/tests/computation/test_eval.py]: add this on to line 20 above

**Review by jreback — CHANGES_REQUESTED** [review]

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
lgtm. ping on green.

**Comment by jreback** [comment]
you have a lint error: https://dev.azure.com/pandas-dev/pandas/_build/results?buildId=5788

File: pandas/core/computation/ops.py [hunk 12]
```diff
@@ -14,6 +15,7 @@
 import pandas as pd
 from pandas.core.base import StringMixin
 import pandas.core.common as com
+from pandas.core.computation.check import _NUMEXPR_INSTALLED, _NUMEXPR_VERSION
```
Comment (jreback) [inline pandas/core/computation/ops.py]: @anjsudh ok, the problem is we can't import here (we have a check in code_checks to avoid importing the optional dependencies until they are needed).

So. I think just add floor & ceil to the _math_ops (always) and better to do a run-time check if this op exists in ne (not exactly sure how to do that).

**Review by jreback — CHANGES_REQUESTED** [review]

**Review by jreback — APPROVED** [review]
